README.md
Imager for Craft CMS 3.x
=====
This plugin is fork from [aelvan/Imager-Craft](https://github.com/aelvan/Imager-Craft/)( DEPRECATED ), I con't promiss timely update it, So if your project is very important, Please use [Imager X](https://plugins.craftcms.com/imager-x).
Imager is a plugin for doing image transforms in Craft templates. It does all the things that the built-in image transform functionality do – but much, much more.

Features
---

- The most efficient template syntax for doing a bunch of image transforms in one go.
- Transforms are completely file based, no database queries needed.
- Transform Assets (local and cloud-based ones), local images, external images, or even the transformed images themselves.
- Upload and serve transforms from Amazon S3 or Google Cloud Storage. Or, write your own storage interface to use whichever cloud service you want.
- Optimize your transformed images with jpegoptim, jpegtran, mozjpeg, optipng, pngquant, gifsicle, TinyPNG, Imagemin or Kraken. Or, write your own optimizer interface for whichever post-optimization tool or service you want.
- Support for offloading all your transforms to Imgix.
- Support for interlaced/progressive images.
- Support for animated gifs.
- In addition to jpeg, gif and png, you can save images in webp format (if you have the necessary server requirements).
- Crop position is relative (in percent) not confined to edges/center (but the built-in keywords still works).
`{ width: 600, height: 600, mode: 'crop', position: '20% 65%' }`
- New cropZoom parameter for when you want to get a little closer.
`{ width: 600, height: 600, mode: 'crop', position: '20% 65%', cropZoom: 1.5 }`
- New croponly mode. To crop, not resize.
`{ width: 600, height: 600, mode: 'croponly', position: '20% 65%' }`
- Easily use Craft's built in focal point when cropping.
`{ width: 600, height: 600, position: asset.getFocalPoint() }`
- New letterbox resize mode.
`{ width: 600, height: 600, mode: 'letterbox', letterbox: { color: '#000', opacity: 0 } }`
- If you know the aspect ratio you want, you don't have to calculate the extra height/width.
`{ width: 800, ratio: 16/9 }`
- Basic image effects, including grayscale, negative, blur, sharpen, gamma and colorize.
`{ effects: { sharpen: true, gamma: 1.4, colorize: '#ff9933' } }`
- Advanced effects, including color blend, tint, sepia, contrast, modulate, normalize, contrast stretch, unsharp mask, posterize and vignette (Imagick imagedriver only).
`{ effects: { modulate: [100, 40, 100], colorBlend: ['rgb(255, 153, 51)', 0.5] } }`
- Your own choice of which resize filter to use. Speed vs. quality is up to you (Imagick imagedriver only).
- Concerned about people copying your images? You can add a watermark to them with Imager.
`{ watermark: { image: logo, width: 80, height: 80, position: { right: 30, bottom: 30 }, opacity: 0.8, blendMode: 'multiply' } }`
- Imager also lets you get color information, dominant color and palette, from your images.
- Imager also includes a bunch of color utilities for getting brightness, hue, lightness, percieved brightness, relative luminance, saturation, brightness difference, color difference and (puh!) contrast ratio.

---

Relevant articles
---
[What's new in Imager 2.0?](https://github.com/DandeSuen/Imager-Craft/wiki/What's-new-in-Imager-2.0%3F)
[Support for Imgix in Imager](https://www.vaersaagod.no/en/support-for-imgix-in-imager-for-craftcms) (Imager 1.x/Craft 2)
[9 tips for speeding up your Imager transforms](https://www.vaersaagod.no/en/9-tips-for-speeding-up-your-imager-transforms-in-craftcms) (Imager 1.x/Craft 2)
[Using WebP images in Craft with Imager](https://www.vaersaagod.no/en/using-webp-images-in-craft-with-imager) (Imager 1.x/Craft 2)

---

Installation
---
To install Imager, follow these steps:

1. Install with composer via `composer require dande/imager` from your project directory.
2. Install the plugin in the Craft Control Panel under Settings > Plugins, or from the command line via `./craft install/plugin imager`.

Imager 2.0 requires Craft 3.x. The Craft 2 version is available in [the master branch](https://github.com/DandeSuen/Imager-Craft/blob/master).

---

Configuration
---

All configuration settings can be overridden by creating an `imager.php` file in your config folder, and adding parameters as needed. Please refer to the [Settings.php file](https://github.com/dande-sun/Imager-Craft/blob/craft3/src/models/Settings.php) for additional pointers regarding available parameters and defaults.

### transformer [string]
*Default: `'craft'`*
*Allowed values: `'craft'`, `'imgix'`*
Transformers is a new concept in Imager 2.0, which provides an overall interface for doing image transforms. There are two built-in transformers at the moment, the default `'craft'` which transforms images on your webserver using the Imagine library that Craft ships with and GD or Imagick, and `'imgix'` which offloads your image transforms to [Imgix](https://www.imgix.com/).

If you want to use Imgix, please refer to the `imgixProfile` and `imgixConfig` config settings, which are required.

### imagerSystemPath [string]
*Default: `'@webroot/imager/'`*
File system path to the folder where you want to store the transformed images.

**Please note:** This folder acts as the cache for transformed images, even if you choose to upload images to some other storage (ie AWS or similar). If you use atomic deploys (ServerPilot, Opworks, or similar), you'd want this folder to be on a shared storage.

### imagerUrl [string]
*Default: `'/imager/'`*
Url to the transformed images. The imagerUrl will be prepended to the path and filename of the transformed image. Can be a relative url, or a full url. If you upload files to AWS, you'd set the imagerUrl to the AWS/CloudFront URL, like so:

    'imagerUrl' => 'http://s3-eu-west-1.amazonaws.com/imagertransforms/',

### cacheEnabled [bool]
*Default: `true`*
Enables or disables caching of transformed images.

### cacheRemoteFiles [bool]
*Default: `true`*
Enables or disables caching of remote files. This includes files on asset sources that are not local, and images on remote URLs.

By default images will be downloaded to your storages folder when first encountered, and cached for the duration of `cacheDurationRemoteFiles`. If you disable it, the downloaded images will be deleted automatically _after each transform request_.

### cacheDuration [int]
*Default: `1209600`*
The cache duration of transformed images.

### cacheDurationRemoteFiles [int]
*Default: `1209600`*
When a remote file is downloaded, it will be cached locally for this duration.

### cacheDurationExternalStorage [int]
*Default: `1209600`*
The cache duration that is set on images uploaded to external storages.

### cacheDurationNonOptimized [int]
*Default: `300`*  
The cache duration that is set on images uploaded to external storages if a post optimization for the image is scheduled. 

### jpegQuality [int]
*Default: `80`*  
Defines the JPEG compression level. Higher values equals better quality and bigger filesizes.  

### pngCompressionLevel [int]
*Default: `2`*  
Defines the PNG compression level. PNG compression is always lossless, so this setting doesn't have any effect on quality. It only affects speed and filesize. A lower value means faster compression, but bigger filesizes. A value of 0 means "no compression", which is the preferred setting if you're doing any post optimizations of png images (with Optipng or TinyPNG).

### webpQuality [int]
*Default: `80`*  
Defines the WEBP compression level. Higher values equals better quality and bigger filesizes.

### webpImagickOptions [array]
*Default: `array()`*  
Additional options you want to pass to Imagick when creating webp files. See [the ImageMagick documentation](http://www.imagemagick.org/script/webp.php) for possible options (although, all are not supported by Imagick).

Example on how to use it in config files:

    'webpImagickOptions' => array(
        'lossless' => 'true',
        'method' => '5',
    ),
    
Example on how to use it in your template code:
    
    {% set webpImage = craft.imager.transformImage(image, { width: 500, format: 'webp', webpImagickOptions: { lossless: 'true', method: '1' } }) %}

### useCwebp [bool]
*Default: `false`*  
If you don't have support for webp in the image driver you're using (GD or Imagick), you can set this to `true` to use the cwebp command line tool instead. cwebp needs to be installed
on the server, but is available on most linux distros.

When using cwebp, Imager will first create a temprary, high quality, transformed image in the source format (jpeg, png or gif), and then convert this to webp.
Since this results in two images being created for each transformed image, the operation is a bit slower (~15-20%) than using the image driver directly. 

### cwebpPath [string]
*Default: `/usr/bin/cwebp`*  
Path to the cwebp binary.

### cwebpOptions [string]
*Default: ``*  
Options to pass o the cwebp binary (use `cwebp -longhelp` to see available options). Quality (-q) is automatically added based on `webpQuality`.  

### interlace [bool|string]
*Default: `false`*  
*Allowed values: `false`, `true` (`'line'`), `'none'`, `'line'`, `'plane'`, `'partition'`*   
Defines the interlace method for creating progressive images. Imagick is required for the `plane` and `partition` methods.
[See demo here](http://imager.vaersaagod.no/?img=5&demo=interlaced) (throttle your connection).  

*GD only supports one interlace mode, so it only makes sense to set a specific interlace mode if you use Imagick.*

### allowUpscale [bool]
*Default: `true`*  
If set to `false`, images will never be upscaled.

### resizeFilter [string]
*Default: `'lanczos'`*  
*Allowed values: `'point'`, `'box'`, `'triangle'`, `'hermite'`, `'hanning'`, `'hamming'`, `'blackman'`, `'gaussian'`, `'quadratic'`, `'cubic'`, `'catrom'`, `'mitchell'`, `'lanczos'`, `'bessel'`, `'sinc'`*  
Sets the resize filter. Imagick is required. By default Craft uses the lanczos filter for interpolation when resizing. It yields good visual result, but is one of the slower ones. There is a bunch of different filters available, and Imager let's you decide which one to use (lanczos is still the default).

The difference in quality and encoding speed varies a lot, so you should choose what is most important for your project, speed or quality. The difference in speed will be more pronounced the bigger the original image is. The difference in quality will be more prononounced the bigger the resulting image is.

Here's the result of a speed test [posted on php.net](http://se1.php.net/manual/en/imagick.resizeimage.php#94493):

```
FILTER_POINT took: 0.334532976151 seconds
FILTER_BOX took: 0.777871131897 seconds
FILTER_TRIANGLE took: 1.3695909977 seconds
FILTER_HERMITE took: 1.35866093636 seconds
FILTER_HANNING took: 4.88722896576 seconds
FILTER_HAMMING took: 4.88665103912 seconds
FILTER_BLACKMAN took: 4.89026689529 seconds
FILTER_GAUSSIAN took: 1.93553304672 seconds
FILTER_QUADRATIC took: 1.93322920799 seconds
FILTER_CUBIC took: 2.58396601677 seconds
FILTER_CATROM took: 2.58508896828 seconds
FILTER_MITCHELL took: 2.58368492126 seconds
FILTER_LANCZOS took: 3.74232912064 seconds
FILTER_BESSEL took: 4.03305602074 seconds
FILTER_SINC took: 4.90098690987 seconds
```

### smartResizeEnabled [bool]
*Default: `false`*  
When set to `true`, the new smartResize method that was added in Craft 2.5 will be used when Imagick is installed. This method is based on [the research done by Dave Newton](http://www.smashingmagazine.com/2015/06/efficient-image-resizing-with-imagemagick/) and yields smaller filesizes without any noticeable loss of quality.

Using smartResize may drastically decrease performance, especially when resizing pngs, and is therefore disabled by default.   

### removeMetadata [bool]
*Default: `false`*   
Strips meta data from image, resulting in smaller images. *Only available with Imagick.*

### bgColor [string]
*Default: `''`*  
If an image is converted from a format that supports transparency, to a format that doesn't, you can specify which background color (in hexadecimal) the resulting image should have.

### position [string]
*Default: `'50% 50%'`*  
By default, Imager will use the native focal point for Asset elements. For other types of source elements, like external images, this setting will be used as default when cropping an image. 

### letterbox [array]
*Default: `array('color'=>'#000', 'opacity'=>0)`*  
Specifies the color and opacity to use for the background when using the letterbox resize method. Opacity is only applicable when saving the transformed file in png format. Animated gifs will always have transparent backgrounds.

### useFilenamePattern [bool]
*Default: `true`*  
When enabled, the path of the transformed asset will be created using the `filenamePattern`.  

### filenamePattern [bool]
*Default: `'{basename}_{transformString|hash}.{extension}'`*  
Pattern that defines how the filename of the transformed file should look. The following variables are available:

**{basename}:** The base name of the original file that was transformed (everything before the extension).  
**{extension}:** The extension of the transformed file.  
**{fullname}:** The full name of the transformed file, basename plus the transform string.  
**{transformString}:** The generated transform string.  

`{basename}`, `{fullname}`, and `{transformString}` can be hashed, either with a long md5 hash, or a shorter, truncated one based the `shortHashLength` config setting.

**Examples:**

    // Show name and transform string (good for debugging
    'filenamePattern' => '{fullname}.{extension}'

    // Show name and a hashed transform string (default)
    'filenamePattern' => '{basename}_{transformString|hash}.{extension}'

    // Show name and a shorter hashed transform string 
    // (prettier, but could result in name collisions (not likely though))
    'filenamePattern' => '{basename}_{transformString|shorthash}.{extension}'

You could use this config setting in your templates to create more SEO-friendly image filenames.

**Example:**

    {% set employeeImage = craft.imager.transformImage(employee.image, { width: 400 }, {}, 
        { filenamePattern: employee.name ~ '_{fullname|shorthash}.{extension}' }) 
    %}

Make sure to always include the transform string in some way or another to avoid transforms getting the same name and overwriting each other.

### shortHashLength [int]
*Default: `10`*  
The length of the short hash when using the `shorthash` filter in `filenamePattern`. 

### hashFilename [bool|string]
*`hashFilename` has been deprecated, use `filenamePattern` instead*
*Default: `'postfix'`*  
*Allowed values: `true`, `false`, `'postfix'`*   
When doing an transform, Imager creates a filename based on the properties in the transform. By default (`'postfix'`), the generated part of the filenamed is hashed, and added as a postfix to the original filename. This makes the filename still relevant for SEO purposes, but still reasonably short.

If set to `false`, the filename will contain the generated string in clear text. This can result in really long filenames, but is great for debugging purposes.

If set to `true`, the whole filename is hashed. This will result in a short, but obscured, filename.

### hashPath [bool]
*Default: `false`*  
When enabled, the path of the transformed asset will be hashed.  

### addVolumeToPath [bool]
*Default: `true`*  
When enabled, the handle for the volume is added to the path of the transformed image. When disabled, transformes from different volumes with the exact same name and transform string could overwrite each other.

### hashRemoteUrl [bool|string]
*Default: `false`*  
*Allowed values: `true`, `false`, `'host'`*   
When tranforming remote images, the hostname and remote path will be used as path inside your imager system path. If you want to shorten or obfuscate the remote url, you can set this to `true` or `'host'`.

If you set this to `true`, the whole url will be hashed and used as the path.

If set to `'host'`, only the hostname will be hashed, while the remote path will be kept. 

### useRemoteUrlQueryString [bool]
*Default: `false`*  
By default, query strings on external urls are not used when creating the filename of the transformed file, to improve caching. When enabled, any query string on the external url will be hashed and added as a part of the filename.  

### instanceReuseEnabled [bool]
*Default: `false`*  
By default, both in Imager and Craft's built in transform functionality, the original image is loaded into memory for every transform. This ensures that the quality of the resulting transform is as good as possible. 

If set to `true`, the original image is only loaded once, and every transform will continue working on the same image instance. This significantly increases performance and memory use, but will most likely decrease quality. See [this demo page](http://imager.vaersaagod.no/?img=5&demo=batch-reuse) for an example of how it works.

### noop [bool]
*Default: `false`*
Setting `noop` (no operation) to `true` makes Imager return the source image untouched. Useful if you don't want to do image transforms in
some environments.

### noopSaveLocal [bool]
*Default: `false`*
If setting `noop` to `true` And Setting `noopSaveLocal` to `true` makes Imager the source save to local (untouched) and return the local image untouched.

### suppressExceptions [bool]
*Default: `false` if devMode is enabled, otherwise `true`*
By default Imager throws exceptions if file based operations fail, an external image can't be downloaded, etc. If `suppressExceptions` is set to `true`, Imager will instead log errors to the log file, and return `null` to the template.

### convertToRGB [bool]
*Default: `false`*
Enable this setting to ensure that transformed images are saved as RGB.

### convertSVG [bool]
*Default: `false`*
Now, Imager default not conver .SVG file ( vector image conver to bitmap image have error, GD not support webp, cwebp not support SVG, Imagick conver SVG with the source image is different ).

### skipExecutableExistCheck [string]
*Default: `false`*
By default Imager will check if the executables for the post-transform optimizations exists. If you have basedir restrictions on, or for some other reason doesn't want to do this check, set this to true.

### curlOptions [array]
*Default: `array()`*
Allows you to add [cURL options](http://php.net/manual/en/function.curl-setopt.php) to be used when downloading images from an external host.

**Example:**

    'curlOptions' => array(
      CURLOPT_SSL_VERIFYPEER => false,
      CURLOPT_SSLVERSION => 1,
      CURLOPT_SSL_CIPHER_LIST => 'TLSv1'
    ),

You can also override the default options, which currently is:

    $defaultOptions = array(
      CURLOPT_HEADER => 0,
      CURLOPT_FOLLOWLOCATION => 1,
      CURLOPT_TIMEOUT => 30
    );   

Make sure you **don't** set `CURLOPT_FILE`, since it is set by Imager to be the output location for the downloaded file.   
  
### runJobsImmediatelyOnAjaxRequests [bool]  
*Default: `true`*  
Craft automatically runs any pending tasks on normal site requests, but not on ajax-request, leaving any optimization jobs that Imager has created in a queue that is run on the next CP request. By default Imager solves this by triggering `queue/run` manually if the request was an ajax request, and a task was created (curl needed). You can disable this behavior by changing this setting to `false`.   
  
### fillTransforms [bool]
*Default: `false`*  
Enable this setting to automatically fill a transform array with additional transforms based on `fillAttribute` and `fillInterval`.    

### fillAttribute [string]
*Default: `'width'`*  
Attribute to be used when filling in the transforms array. Can be any valid numeric attribute.
     
### fillInterval [string]
*Default: `200`*  
Interval to be used when filling in the transforms array. This should always be a positive integer, Imager will automatically figure out if
the transform has been ordered in an ascending or descending order.

### clearKey [string]  
*Default: `''`*  
Key to use when clearing the transform or remote images cache with the controller actions. An empty string means clearing is disabled.

### useForNativeTransforms [bool]  
*Default: `false`*  
Overrides Craft's internal transform functionality with Imager's. When enabled, `myAsset.getUrl({ width: 400 })` will result in an Imager transform.

### useForCpThumbs [bool]  
*Default: `false`*  
Overrides Craft's internal thumb transform in the control panel with Imager's. Mostly a good idea if you're using Imgix, if not you'd probably be better off not enabling this.

### imgixProfile [string]  
*Default: `'default'`*  
Imgix config settings profile to be used. Must correspond to a key in `imgixConfig`.

### imgixApiKey [string]
*Default: `''`*  
A valid Imgix API key is required to enable the [Imgix purging](https://docs.imgix.com/setup/purging-images) features. Note that in addition to setting the `imgixApiKey` setting, you can also add per-profile API keys for specific sources inside the `imgixConfig` configuration object – useful if you're using sources belonging to different Imgix accounts.

### imgixEnableAutoPurging [bool]
*Default: `true`*  
Attempts to [purge](https://docs.imgix.com/setup/purging-images) images automatically when the Asset is replaced, or edited with the Image Editor (i.e. cropped, resized etc). _An Imgix API key is required to enable purging._  

### imgixEnablePurgeElementAction [bool]
*Default: `true`*  
Adds a "Purge from Imgix" element action to the Asset index, for manually triggering purge. Note that if purging is not possible (i.e. if there's no Imgix API key set or there are no purgable profiles in the `imgixConfig` array) the element action will not display, regardless of this setting.  

### imgixConfig [array]  
An array of configuration objects for Imgix, where the key is the profile handle. The configuration object takes the following settings:

**domains (array):** An array of Imgix source domains.  

**useHttps (bool):** Indicates if generated Imgix URLs should be https or not.  

**signKey (string):** If you've protected your source with secure URLs, you must provide the sign key/token. An empty string indicates that the source is not secure.  

**sourceIsWebProxy (bool):** Indicates if your Imgix source is a web proxy or not. Note that web proxy sources will be excluded from purging.  

**useCloudSourcePath (bool):** If enabled, Imager will prepend the Craft source path to the asset path, before passing it to the Imgix URL builder. This makes it possible to have one Imgix source pulling images from many Craft volumes when they are on the same S3 bucket, but in different subfolder. This only works on volumes that implements a path setting (AWS S3 and GCS does, local volumes does not).  

**addPath (string|array):** Prepends a path to the asset's path. Can be useful if you have several volumes that you want to serve with one Imgix web folder source. If this setting is an array, the key should be the volume handle, and the value the path to add. See example below.  

**shardStrategy (string):** etermines the sharding strategy if more than one source is used. Allowed values are `cycle` and `crc`.  

**getExternalImageDimensions (bool):** Imager does its best at determining the dimensions of the transformed images. If the supplied asset is on Craft source, it's easy because Craft records the original dimensions of the image in the database. But if the image is external, it's not that easy. Imager will try to determine the size based on the transform parameters, and if both width and height, or ratio is provided, it'll usually be able to. But if you only transform by one attribute, it may not be possible. In these cases Imager will by default download the source image and check the dimensions to calculate the missing bits.

By disabling this setting, you're telling Imager to never download external images, and to just give up on trying to figure out the dimensions. If you supplied only width to the transform, height will then be set to 0. If you don't need to use height in your code, that's totally fine, and you've managed to squeeze out a bit more performance.


CHANGELOG.md
# Imager Changelog

## 2.4.3.4 - 2020-09-25

- Modified: use `Twig\TwigFilter` instead `Twig_SimpleFilter`.
- Modified: use `yii\base\InvalidArgumentException` instead `yii\base\InvalidParamException`.

## 2.4.3.2 - 2020-09-24

- Modified fillter `srcsetAttr` ( Support IE, Lazysize, placeholder )

## 2.4.3.1 - 2020-09-23

- Fixed `composer require dande/imager` not working.

## 2.4.3 - 2020-09-22

- Add Azure blog Storage support ( Now, The Imager-craft support upload to Azure blog storage. )
- Add fillter `srcsetAttr` ( Auto output html tag attribute, if input is array, output `srcset="url1 width, url2 width"` )
- Add fillter `srcsetCustomAttr`
- Add fillter `silhouette`
- Add new config option `convertSVG`, Default value `false`. ( `true` convert svg, `false` not convert svg )
- Add new config option `noopSaveLocal`, Default value `true`. ( noop == true && noopSaveLocal == true , Imager will copy source file to local )

## 2.4.2 - 2020-06-11

- Change new version number.

## 2.4.1 - 2020-06-11

- Filtered unkown suffix ( cantains .svg )

## 2.4.0 - 2020-02-11

> {warning} This version of Imager is no longer actively maintained. [Imager X](https://plugins.craftcms.com/imager-x) is the successor of Imager 2.0, a commercial plugin with several new and awesome features.

### Changed
- Changed README to inform about [Imager X](https://plugins.craftcms.com/imager-x).  

### Fixed
- Fixed an issue where Imgix URLs with special characters were being encoded unnecessarily. 

## 2.3.1 - 2019-12-30
### Fixed
- Fixed an issue where realpath() was used unnecessarily and introduced a bug that would result in incorrect file paths when uploading to external storages (fixes #289, thanks @SanderVanLeeuwen).

## 2.3.0 - 2019-09-21

> {warning} As of Imager 2.3.0, the native focal point will automatically be used for transforms on Asset elements, _if no `position` is set in the transform_. If you were relying on the configuration setting `position` as the fallback value in previous versions, this could result in new transforms being created for some images, and you should consider adding your default value explicitly before upgrading. 

### Changed
- Imager will now automatically use native focal point for Asset elements, if no position is set by the transform. (fixes #270). 


## 2.2.0 - 2019-09-16
### Added
- Added missing `noop` implementation (fixes #260). 

### Changed
- Changed how watermark transform string is generated to avoid collisions due to keys in position (fixes #244). 
- AWS credentials can now be configured via IAM role, shared credentials file, or environment variables, when `accessKey` and `secretAccessKey` are left blank in the storage config settings (Thanks, @carlcs). 
- The `requestHeaders` and `storageType` storage config settings for AWS are now optional (Thanks, @carlcs).

### Fixed
- Fixed an issue where an error thrown while purging images on Imgix could result in assets not being replaced correctly and JS alerts IN CP. Purging is now silent, check logs for errors (fixes #261).
- Fixed issue where gif placeholder would not work if the image driver was left at `'auto'` in Craft's general config (fixes #195).
- Fixed incorrect calculation of image size in `ImgixTransformedImageModel` under very specific circumstances (fixes #238).
- Normalized path to `imagerSystemPath` returned from `Settings` model (Thanks, @andersaloof).
- Fixed directory separators in temp file path.

## 2.1.10 - 2019-04-13
### Fixed
- Fixed issues with using named transforms or `AssetTransform` models when `useForNativeTransforms` is `true` (fixes #237).
- Fixed incorrect use of config setting `imagerSystemPath` when clearing cache (fixes #236).
- Fixed docs for imgixConfig, domains are suppose to be strings.
- Fixed default value for `getExternalImageDimensions` in `ImgixSettings` model to correspond with the description in docs, and the example in the `Settings` model.

## 2.1.9 - 2019-04-04
### Fixed
- Fixed path encoding for AWS and GCS when using Windows (Thanks, @aaronwaldon).
- Fixed path encoding for Imgix when using Windows (Thanks, @aaronwaldon).
- Fixed error that would occur when using cwebp and the path had a space. (fixes #220).
- Optimize job now throws an exception if upload to external storage fails, so that it can be retried.
- AWS and GCS external storages now open files in binary mode just to be safe.

## 2.1.8.1 - 2019-04-03
### Added
- Added environment parsing to volume subfolder setting in Imgix path parsing (Thanks, @Tam).

## 2.1.8 - 2019-04-02
### Added
- Added support for addPath in Imgix config to make it easier to use several volumes with just one Imgix source and configuration.

## 2.1.7 - 2019-03-20
### Fixed
- Fixed regression error where settings that could have alias was not possible to override with template config overrides.

## 2.1.6 - 2019-03-19
### Fixed
- Fixes issue where spaces in asset folders broke URL encoding for Imgix.
- Fixes missing width and height when using Imgix and neither width or height was set in the transform (fixes #228).

## 2.1.5.1 - 2019-03-08
### Fixed
- Fixes issue where files with upper case extensions were not transformed with Imager when useForCpThumbs was activated. (thanks, @johandouma).

## 2.1.5 - 2019-02-21
### Fixed
- Fixes issue where files with upper case extensions were not transformed with Imager when useForNativeTransforms was activated. (thanks, @johandouma).

## 2.1.4 - 2019-01-08
### Fixed
- Makes URL encoding of file paths for Imgix RFC 3986 compliant (#190) (thanks, @Mosnar).
- Fixes in issue with running Imager transforms from the command line (thanks, @janhenckens).

## 2.1.3 - 2018-12-31
### Fixed
- Default quality is no longer sent to Imgix when auto compression is enabled (thanks, @jorenvanhee).

## 2.1.2 - 2018-11-27
### Added
- Added support for purging images from Imgix (thanks, @mmikkel).

## 2.1.1 - 2018-11-01
### Fixed

src/models/Settings.php
<?php

namespace aelvan\imager\models;

use craft\helpers\FileHelper;
use craft\base\Model;
use Yii;

class Settings extends Model
{
    public $transformer = 'craft';
    public $imagerSystemPath = '@webroot/imager/';
    public $imagerUrl = '/imager/';
    public $cacheEnabled = true;
    public $cacheRemoteFiles = true;
    public $cacheDuration = 1209600;
    public $cacheDurationRemoteFiles = 1209600;
    public $cacheDurationExternalStorage = 1209600;
    public $cacheDurationNonOptimized = 300;
    public $jpegQuality = 80;
    public $pngCompressionLevel = 2;
    public $webpQuality = 80;
    public $webpImagickOptions = [];
    public $useCwebp = false;
    public $cwebpPath = '/usr/bin/cwebp';
    public $cwebpOptions = '';
    public $interlace = false;
    public $allowUpscale = true;
    public $resizeFilter = 'lanczos';
    public $smartResizeEnabled = false;
    public $removeMetadata = false;
    public $bgColor = '';
    public $position = '50% 50%';
    public $letterbox = ['color' => '#000', 'opacity' => 0];
    public $useFilenamePattern = true;
    public $filenamePattern = '{basename}_{transformString|hash}.{extension}';
    public $shortHashLength = 10;
    public $hashFilename = 'postfix'; // deprecated
    public $hashPath = false;
    public $addVolumeToPath = true;
    public $hashRemoteUrl = false;
    public $useRemoteUrlQueryString = false;
    public $instanceReuseEnabled = false;
    public $noop = false;
    public $noopSaveLocal = true;
    public $suppressExceptions = false;
    public $convertToRGB = false;
    public $skipExecutableExistCheck = false;
    public $curlOptions = [];
    public $runJobsImmediatelyOnAjaxRequests = true;
    public $fillTransforms = false;
    public $fillAttribute = 'width';
    public $fillInterval = '200';
    public $clearKey = '';
    public $convertSVG = false;
    public $useForNativeTransforms = false;
    public $useForCpThumbs = false;

    public $imgixProfile = 'default';
    public $imgixApiKey = '';
    public $imgixEnableAutoPurging = true;
    public $imgixEnablePurgeElementAction = true;
    public $imgixConfig = [
        'default' => [
            'domains' => [],
            'useHttps' => true,
            'signKey' => '',
            'sourceIsWebProxy' => false,
            'useCloudSourcePath' => true,
            'shardStrategy' => 'cycle',
            'getExternalImageDimensions' => true,
            'defaultParams' => [],
            'apiKey' => '',
            'excludeFromPurge' => false,
        ]
    ];

    public $optimizeType = 'job';
    public $optimizers = [];
    public $optimizerConfig = [
        'jpegoptim' => [
            'extensions' => ['jpg'],
            'path' => '/usr/bin/jpegoptim',
            'optionString' => '-s',
        ],
        'jpegtran' => [
            'extensions' => ['jpg'],
            'path' => '/usr/bin/jpegtran',
            'optionString' => '-optimize -copy none',
        ],
        'mozjpeg' => [
            'extensions' => ['jpg'],
            'path' => '/usr/bin/mozjpeg',
            'optionString' => '-optimize -copy none',
        ],
        'optipng' => [
            'extensions' => ['png'],
            'path' => '/usr/bin/optipng',
            'optionString' => '-o2',
        ],
        'pngquant' => [
            'extensions' => ['png'],
            'path' => '/usr/bin/pngquant',
            'optionString' => '--strip --skip-if-larger',
        ],
        'gifsicle' => [
            'extensions' => ['gif'],
            'path' => '/usr/bin/gifsicle',
            'optionString' => '--optimize=3 --colors 256',
        ],
        'tinypng' => [
            'extensions' => ['png','jpg'],
            'apiKey' => '',
        ],
        'kraken' => [
            'extensions' => ['png', 'jpg', 'gif'],
            'apiKey' => '',
            'apiSecret' => '',
            'additionalParams' => [
                'lossy' => true,
            ]
        ],
        'imageoptim' => [
            'extensions' => ['png', 'jpg', 'gif'],
            'apiUsername' => '',
            'quality' => 'medium'
        ],
    ];

    public $srcsetConfig = [
      'srcDefault'  => 'max', // ['min', 'middle', 'max', 'first', 'center', 'last'] or Number
      'PHType'      => 'silhouette', // ['silhouette','placeholder'] or null
      'PHConfig'    => [],
      'descriptor'  => 'w', // ['w', 'h', 'x+h', 'vw', 'vh', 'x']
      'prefix'      => ''
    ];

    public $storages = [];
    public $storageConfig = [
        'aws' => [
            'accessKey' => '',
            'secretAccessKey' => '',
            'region' => '',
            'bucket' => '',
            'folder' => '',
            'requestHeaders' => array(),
            'storageType' => 'standard',
            'cloudfrontInvalidateEnabled' => false,
            'cloudfrontDistributionId' => '',
        ],
        'gcs' => [
            'keyFile' => '',
            'bucket' => '',
            'folder' => '',
        ],
        'abs' => [
          'useHttps'      => true,
          'accountName'   => '',
          'accountKey'    => '',
          'containerName' => ''
        ]
    ];

    /**
     * Settings constructor.
     *
     * @param array $config
     */
    public function __construct($config = [])
    {
        parent::__construct($config);

        if (!empty($config)) {
            \Yii::configure($this, $config);
        }
        $this->init();
    }

    public function init()
    {
        // Have to set this here cause Yii::getAlias can't be used in default value
        $this->imagerSystemPath = FileHelper::normalizePath(Yii::getAlias($this->imagerSystemPath));
        $this->suppressExceptions = !\Craft::$app->getConfig()->general->devMode;
    }
}
